activerecord/lib/active_record/reflection.rb

```diff
@@ -364,6 +364,17 @@ def initialize(name, scope, options, active_record)
         @constructable = calculate_constructable(macro, options)
         @association_scope_cache = {}
         @scope_lock = Mutex.new
+
+        if options[:class_name] && options[:class_name].class == Class
+          ActiveSupport::Deprecation.warn(<<-MSG.squish)
+            Passing a class to the `class_name` option of reflection
+            is eagerloading more classes than necessary and potentually
+            creating circular dependencies.
+
+            Please pass the class name as a string.
+            Example: `belongs_to :client, class_name: 'Company'`
```

Comment:
I agree about talking about the hard error in the next release but I don't think we should mention about `require` even more that require in app is a anti-pattern.